### DISCUSSION TITLE
How to smoothly change speed of an object?

### DISCUSSION THREAD

**Discussion by NikRetaNCAM** [HaxeFlixel/flixel#2554] [Q&A] [answered]
Every object takes time to slow down right?
![image](https://user-images.githubusercontent.com/60806448/167996495-a29d0372-c286-48ac-81c3-9d86f29d4715.png)
this code IMMEDIATELY stops the object, but i want it to take a bit of time to slow down

**Comment by NikRetaNCAM** [comment]
here's what i use for acceleration by the way
![image](https://user-images.githubusercontent.com/60806448/167996597-ca548638-a342-4c31-9525-c7cc3dd04e8d.png)

**Comment by Geokureli** [comment, marked as answer]
rather than setting an object's velocity directly you should use acceleration and drag. I see you're just setting a speed value, I assume that sets velocity, instead set acceleration, and cap it with maxVelocity

this project calculates acceleration and drag based on how long it should take to speed up to max speed and how long to come to a full stop (Source linked in the page)
https://www.newgrounds.com/portal/view/723552

Here's the game's actual code to solve for maxVelocity, acceleration and drag. Note that maxSpeed is determined by how far the player is supposed to be able to jump.
https://github.com/Geokureli/Dial-A-Platformer/blob/master/source/art/DialAPlatformer.hx#L242-L265

**Reply by NikRetaNCAM** [reply]
> rather than setting an object's velocity directly you should use acceleration and drag. I see you're just setting a speed value, I assume that sets velocity, instead set acceleration, and cap it with maxVelocity
> 
> this project calculates acceleration and drag based on how long it should take to speed up to max speed and how long to come to a full stop (Source linked in the page) https://www.newgrounds.com/portal/view/723552
> 
> Here's the game's actual code to solve for maxVelocity, acceleration and drag. Note that maxSpeed is determined by how far the player is supposed to be able to jump. https://github.com/Geokureli/Dial-A-Platformer/blob/master/source/art/DialAPlatformer.hx#L242-L265

i know how ot use acceleration for gravity, can you correct me on how it works? i suppose it speeds object over time due to the name, right? so techncially i should do:
-Detect if key is pressed
-Set acceleration if it's pressed
-Then set.. Minus/opposite acceleration when key is released?

(idk if that''ll help, but it's code for Sonic actually, if you played it you might get how Sonic accelerates and slows down)

**Reply by Geokureli** [reply]
maxVelocity and drag only need to be set once
`maxVelocity` prevents the sprites velocity from increasing past that in either direction, meaning a `maxVelocity.x` of 100 will keep `velocity.x` between -100 and 100

`acceleration` is added to velocity, if `acceleration` is 100, initial `velocity` is 0 and `maxVelocity` is 200 then it will take 2 seconds to reach `maxVelocity`

`drag` is a dampening force applied to `velocity` whenever acceleration is 0. so if drag is 100 and velocity is 100 the object will stop in one second.

As an example, if you set `maxVelocity.x` to 100, and `drag.x` to 200, and set `acceleration.x` to 200 when pressing `RIGHT` and 0 when released, you'll have a player that takes .5s to reach top speed, and .5s to slow down

Check out the [EZ Platformer Demo](https://haxeflixel.com/demos/EZPlatformer/) for another example.
[PlayState.hx#L95-L101](https://github.com/HaxeFlixel/flixel-demos/blob/dev/Platformers/EZPlatformer/source/PlayState.hx#L95-L101)
```hx
// Create _player
_player = new FlxSprite(FlxG.width / 2 - 5);
_player.makeGraphic(8, 8, FlxColor.RED);
_player.maxVelocity.set(80, 200);
_player.acceleration.y = 200;
_player.drag.x = _player.maxVelocity.x * 4;
add(_player);
```
and in update: [PlayState.hx#L122-L130](https://github.com/HaxeFlixel/flixel-demos/blob/dev/Platformers/EZPlatformer/source/PlayState.hx#L122-L130)
```hx
if (FlxG.keys.anyPressed([LEFT, A]))
{
	_player.acceleration.x = -_player.maxVelocity.x * 4;
}

if (FlxG.keys.anyPressed([RIGHT, D]))
{
	_player.acceleration.x = _player.maxVelocity.x * 4;
}
```

By setting `acceleration.x` and `drag.x` to `maxVelocity.x * 4` it means "speed up to and slow from max speed in 1/4th of a second"

**Reply by NikRetaNCAM** [reply]
> 

oh, that's useful. acceleration affects velocity, right? so technically by checking what velocity is object at right now, i can apply correct animations, am i right?

**Reply by Geokureli** [reply]
sure

**Reply by NikRetaNCAM** [reply]
> sure

cool, um.. another question. what EXACTLY does drag do? like, the bigger the drag, the faster object will slow down, or opposite?

**Reply by Geokureli** [reply]
I answered that above
> `drag` is a dampening force applied to `velocity` whenever acceleration is 0. so if drag is 100 and velocity is 100 the object will stop in one second.

A positive drag value will always slow velocity to zero, higher drag means it will slow down quicker

**Reply by NikRetaNCAM** [reply]
wow, this works so much better in so many levels! thank you, Geokureli!
